rfcs/build-system/01_convergence_devstack.md

```diff
@@ -0,0 +1,191 @@
+# WIP | FUI Convergence - devstack/DX
+
+> **tags:** pitch, convergence
+
+> For instructions on shaping a project see here: [Shaping a Project](/kX02SXVbS6KzMOQd56i6Cg)
+>
+> Ported from [hackmd](https://hackmd.io/MiS9ISfjRXqmd6XmxpyMaA)
+
+### Problem
+
+Based on [high level repo issues notes](https://hackmd.io/0TIUjU8_T1Ga2Wjez4IiSQ), we need to address most crucial ones sooner than later to enable good enough DX for both inside/outside contributors.
+
+> Good enough DX:
+>
+> - rapid feedback loop
+> - one common way how to do/run things
+> - docs (reflected within tooling -> `yarn test --help` should just work)
+
+### Appetite
+
+As an initial step lets focus on resolve following (smallest possible area):
+
+1. need to build before run/test/demo
+1. unified TSConfigs
+1. unified way how to transpile javascript
+1. solution for convergence/whole repo
+
+### Solution
+
+> show don't tell / TL;DR
+> ![](https://i.imgur.com/bDwWu9v.png)
+
+#### 1. need to build before run/test/lint/demo
+
+Current state:
+
+- N\* uses wepback aliases with lerna-alias
+- OUIF uses yarn workspaces heavily (symlinks of build assets), thus after every modification outside the package that is being worked on, you need to manually rebuild to get latest changes.
+
+Convergence proposal:
+
+- define one source of truth for aliasing packages that can be leveraged consistently in all tooling (bundlers,testing,demos/storybook)
+
+**👉 tsconfig path aliases + TS solution style configs**
```

Comment:
One extra advantage to using `index.ts` in `main` is that we can remove all jest and webpack aliases, which will still be required if we use TS path aliases

Reply:
> One extra advantage to using index.ts in main is that we can remove all jest and webpack aliases, which will still be required if we use TS path aliases
I forgot that using `main` means to continue to use symlinks (lerna/yarn workspaces) which has various downsides...  

---

 I should have created 4 proposals instead of 1 that tackles "all current dev stack issues" 🥶 

What would be very helpful, is to prepare counter arguments/other solutions in a separate doc with proper reasoning (historical reasoning as well if needed) about:
- Why do we need TS project refs?
- What other approaches did we try and might be a good solution as well?

Otherwise I feel like I'm being in a constant loop without getting to common conclusion.

Reply:
To be clear, I'm happy with any improvement to the current situation :-). I checked out the material ui repo and see that they also use a similar setup to the one you are proposing. 

I just think that having project refs is an improvement, this is what I am pointing out.

Maybe we should define what it means to come to a common conclusion :-)

Reply:
> I forgot that using main means to continue to use symlinks (lerna/yarn workspaces) which has various downsides...
> Why do we need TS project refs?

From my experience with this repo and an internal repo: with projects references with `main` entries I had better experience with IDE, it jumps to source files when you're trying to use "Go to a definition".

In this repo it in 99% jumps to a `.d.ts` file or to a symlink in `node_modules/@fluentui/XXX` 😭 

P.S. I am using Webstorm

Reply:
@layershifter yup that's suboptimal DX, I agree.

 - that's why I'd suggest to stop using lerna/yarn workspaces all over the place - achievable via TS path aliases (project refs can be still be leveraged for build time perf of ambient definitions )
 - to be able to use `main` hack, we would still need to rely heavily on lerna/yarn workspaces

> P.S. I am using Webstorm

same happens in vscode 🔥

Reply:
> * that's why I'd suggest to stop using lerna/yarn workspaces all over the place - achievable via TS path aliases (project refs can be still be leveraged for build time perf of ambient definitions )

What does would be the end to end impact of "stop using lerna/yarn workspaces"? Would it still be possible to run scripts within individual projects when appropriate, or would all scripts be defined only at the root?

Reply:
> What would be the end to end impact of "stop using lerna/yarn workspaces"? 

I mentioned this within #16706 - single version policy for devTooling with possibility to opt out if _IT IS REALLY NECESSARY_ (I don't see such a need especially for convergence/design system within monorepo). To "correct what I said", workspaces can be used for executing commands via `yarn workspace` or if not `yarn --cwd` can be leveraged.

Side note: Ideal state from my experience is not to pollute all packages with scripts - for this we would need solid tooling (nx dev tools has this for example) that would unify API surface for every package.

> Would it still be possible to run scripts within individual projects when appropriate, or would all scripts be defined only at the root?

Yes it will, via `yarn --cwd` (for non workspace project)